pkg/cleaner/virtualservices.go

```diff
@@ -0,0 +1,59 @@
+package cleaner
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/giantswarm/microerror"
+	"github.com/go-logr/logr"
+	"github.com/vmware/cloud-provider-for-cloud-director/pkg/vcdsdk"
+	capvcd "github.com/vmware/cluster-api-provider-cloud-director/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type VirtualServiceCleaner struct {
+	cli client.Client
+}
+
+func NewVirtualServiceCleaner(cli client.Client) *VirtualServiceCleaner {
+	return &VirtualServiceCleaner{cli: cli}
+}
+
+// force implementing Cleaner interface
+var _ Cleaner = &VirtualServiceCleaner{}
+
+func (lbc *VirtualServiceCleaner) Clean(ctx context.Context, log logr.Logger, vcdClient *vcdsdk.Client, c *capvcd.VCDCluster) (bool, error) {
+	log = log.WithName("VirtualServiceCleaner")
+	gateway, err := vcdsdk.NewGatewayManager(ctx, vcdClient, c.Spec.OvdcNetwork, c.Spec.LoadBalancerConfigSpec.VipSubnet)
+	if err != nil {
+		return false, err
+	}
+	vSvcs, err := vcdClient.VCDClient.GetAllAlbVirtualServices(gateway.GatewayRef.Id, nil)
+	if err != nil {
+		return false, err
+	}
+	infraId := c.Status.InfraId
+	if len(infraId) == 0 {
+		return true, microerror.Mask(fmt.Errorf(".status.infraId is not populated on the cluster: %s", c.Name))
+	}
+	deleted := 0
+	for _, vSvc := range vSvcs {
+		svcName := vSvc.NsxtAlbVirtualService.Name
+		//todo: pools
+		// if the name of the virtual service contains the cluster's infraId, delete this virtual service
+		if strings.Contains(svcName, infraId) {
+			log.Info(fmt.Sprintf("deleting virtual service: %s", svcName))
+			err = gateway.DeleteVirtualService(ctx, svcName, false)
```

Comment:
Does it work in an async way? Can we assume the LB is deleted when this call returns? If so, can we return true if all LBs are deleted correctly without waiting for next reconciliation loop?

Reply:
the false param here means "don't fail if it's not there", but the whole clean function returns fals in the "happy path scenario" so it's not being re-queued.

y, I think these calls are async, but will double check. Yesterday, I had some troubles to create a healthy vcdcluster on glados so it was hard to check from debugger, but the vServices got cleaned